### PR TITLE
132: add transition_to_archive

### LINKED ISSUE #132: Support transition_to_archive lifecycle policy
### Describe the Feature

AWS recently added a [new storage class to EFS called Archive](https://aws.amazon.com/blogs/aws/optimize-your-storage-costs-for-rarely-accessed-files-with-amazon-efs-archive/)
AWS Terraform provider will soon have a release that adds support to it (see [PR](https://github.com/hashicorp/terraform-provider-aws/pull/35096))

### Expected Behavior

Similar to the existing `transition_to_ia` input, add a new `transition_to_archive` input

### Use Case

The new Archive EFS storage class has potential savings of about 50% of the unused data

### Describe Ideal Solution

Add a new `transition_to_archive` input that supports the relevant values

### Alternatives Considered

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by jgournet**
## what
Add transition_to_archive, same as transition_to_ia
Also add 180, 270 and 365 days options

## why
See https://github.com/cloudposse/terraform-aws-efs/issues/132

## references
closes #132

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -33,6 +33,13 @@ resource "aws_efs_file_system" "default" {
     }
   }
 
+    dynamic "lifecycle_policy" {
```
Comment (Gowiem) [inline main.tf:36]: Formatting issue. Please run `terraform fmt` locally.

**Review by Gowiem — CHANGES_REQUESTED** [review]
@jgournet this looks good. I'm not sure why the originaly implementor chose to do these types of vars as a `list(string)`, but I can see that you're just copying their pattern so I can't fault you for that!

Please run a tf fmt and then also run our Makefile automation so that we can regenerate the README. Please run the following locally and commit + push the result: 

```
make init
make readme
```

Thanks!

**Comment by jgournet** [comment]
Done: ran terraform fmt, make init and make readme
Thanks @Gowiem  !
(and yes, having a list was weird too, but it's my first PR with cloudposse ... so definitely not my place to change things :D )

Also: the automated tests fail to startup with this error:
```
Invalid workflow file: .github/workflows/release-branch.yml#L20
The workflow is not valid. .github/workflows/release-branch.yml (Line: 20, Col: 11): Secret REPO_ACCESS_TOKEN is required, but not provided while calling. .github/workflows/release-branch.yml (Line: 22, Col: 28): Invalid secret, github_access_token is not defined in the referenced workflow.
```
Am I meant to do something about it ?

File: main.tf (hunk 1 above)
Comment (jgournet) [reply main.tf]: thanks - done !

**Comment by Gowiem** [comment]
Aha we're still hitting a formatting issue, but it's in examples/complete. Can you run `tf fmt` in that dir as well and commit the result?

**Comment by jgournet** [comment]
Pushed the format for examples/complete - thanks for the guidance (and sorry for the time this is taking)
note: but tests are still not running

**Comment by Gowiem** [comment]
@jgournet we're getting the following error: 

```
TestExamplesComplete 2024-04-18T21:41:22Z retry.go:99: Returning due to fatal error: FatalError{Underlying: error while running command: exit status 1; ╷
│ Error: Unsupported argument
│ 
│   on ../../main.tf line 39, in resource "aws_efs_file_system" "default":
│   39:       transition_to_archive = try(var.transition_to_archive[0], null)
│ 
│ An argument named "transition_to_archive" is not expected here.
╵}
    apply.go:15: 
        	Error Trace:	apply.go:15
        	            				examples_complete_test.go:37
        	Error:      	Received unexpected error:
        	            	FatalError{Underlying: error while running command: exit status 1; ╷
        	            	│ Error: Unsupported argument
        	            	│ 
        	            	│   on ../../main.tf line 39, in resource "aws_efs_file_system" "default":
        	            	│   39:       transition_to_archive = try(var.transition_to_archive[0], null)
        	            	│ 
        	            	│ An argument named "transition_to_archive" is not expected here.
        	            	╵}
        	Test:       	TestExamplesComplete
```

This is likely due to an issue with the wrong provider version. Can you check in on the provider version where this argument was introduced and update our version pinnings to enforce that we're using that version? Be sure to run make init / readme following you changes as that will create changes in the README. Thanks!

**Comment by Gowiem** [comment]
@jgournet failure in the examples/complete again: 

```
Downloading registry.terraform.io/cloudposse/route53-cluster-hostname/aws 0.13.0 for efs.dns...
╷
│ Error: Failed to query available provider packages
│ 
│ Could not retrieve the list of available versions for provider
│ hashicorp/aws: no available releases match the given constraints >= 2.0.0,
│ >= 3.0.0, >= 3.71.0, >= 4.9.0, 4.42.0, >= 5.32.0
╵
```

Sorry, but mind looking into that?

**Comment by jgournet** [comment]
```
 │ Error: putting EFS file system (fs-0880fb02c3829e333) lifecycle configuration: BadRequest: The ThroughputMode value for the file system does not support TransitionToArchive. Either change the ThroughputMode value to Elastic or remove the TransitionToArchive parameter.
│ {
│   RespMetadata: {
│     StatusCode: 400,
│     RequestID: "9a99c3ab-4c41-4691-ae42-0a670a3a2c40"
│   },
│   ErrorCode: "BadRequest",
│   Message_: "The ThroughputMode value for the file system does not support TransitionToArchive. Either change the ThroughputMode value to Elastic or remove the TransitionToArchive parameter."
│ }
```
checking this one

**Comment by Gowiem** [comment]
That seems like a legit error, but it'd be a painful thing to check prior to runtime @jgournet. Mind changing the examples thoughtput mode and we'll get the tests passing that way?

**Comment by jgournet** [comment]
@Gowiem : I actually removed (with comment) the "transition_to_archive" from the complete example.
=> do you prefer me changing the complete example throughput mode as you mentioned ?

**Comment by Gowiem** [comment]
No your change was fine @jgournet -- Let's see if they pass this time around 👍

**Review by Gowiem — APPROVED** [review]

**Comment by Gowiem** [comment]
@jgournet our release workflow for this repo is breaking for some reason... apologies, but the release to the registry will be delayed while we get that sorted out. I'll try to circle back when fixed. Thanks for the hard work on this PR and the back and forth to get it over the line!

**Comment by jgournet** [comment]
Thank you for your help and patience; and guidance !! really appreciated

**Comment by aballesta** [comment]
Hello @Gowiem , I hope you are doing fine, Is there any progress in fixing the release workflow? the variable transition_to_archive not exist on v1.1.0 and I was wondering when I could use this feature. Thank you.

**Comment by Gowiem** [comment]
@aballesta thanks for the ping. I'll try to look into this when I get the chance. It has been a while for sure so thanks for the patience. I'll try to follow up shortly!

**Pull request merged**